### PR TITLE
[trainer] --model_parallel hasn't been implemented for most models

### LINKED ISSUE #9336: "RuntimeError: Input, output and indices must be on the current device" when trying to finetune MBart
### Environment info
- Platform: Linux-4.15.0-123-generic-x86_64-with-glibc2.10
- Tried `transformers` versions 4.1.1 (installed with pip) and 4.2.2 (installed from master branch of the repository)
- Python version: 3.7
- PyTorch version: 1.7
- Tensorflow version: 2.4
- Number of available GPU: 2 (GeForce RTX 2080 Ti, with ~11GB of memory each)

### Information
Model I am using (Bert, XLNet ...): MBart -> facebook/mbart-large-cc25

The problem arises when using: the official example scripts: (details below)

The tasks I am working on is: my own task or dataset: (details below)

I am fine-tuning MBart using my own dataset, using the `examples/seq2seq/finetune.sh` script. When I run it on a single GPU, I get a memory error, as one GPU has not enough memory to load the MBart model. When I try to distribute the model on two GPUs, I get a RuntimeError: 
`RuntimeError: Input, output and indices must be on the current device`

### To reproduce
I am running the script in the following way:
`CUDA_VISIBLE_DEVICES=0,1 transformers/examples/seq2seq/finetune.sh --model_name_or_path "facebook/mbart-large-cc25" --output_dir output --data_dir data --overwrite_output_dir --model_parallel --per_device_train_batch_size 1 --per_device_eval_batch_size 1 --freeze_encoder --freeze_embeds --tgt_lang "en"`

I have also tried:
`CUDA_VISIBLE_DEVICES=0,1 transformers/examples/seq2seq/finetune.sh --model_name_or_path "facebook/mbart-large-cc25" --output_dir output --data_dir data --overwrite_output_dir --model_parallel --tgt_lang "en"`

I also tried limiting the length of source and target sentences by trying several values for `--max_target_length` and `--max_source_length'`. In addition, I tried using more GPUs (up to 4).

If I run `wc -l` on my `data` directory, I get:
```
3004 data/test.source
3004 data/test.target
686623 data/train.source
686623 data/train.target
2999 data/val.source
2999 data/val.target
```

### PR DESCRIPTION
**Pull request by stas00**
Apparently we unleashed `--model_parallel` in trainer w/o checking if the model supports MP (most don't). This PR:

* [x] checks whether the model supports MP and asserts otherwise
* [x] fixes the cl arg help to note that the flag will only work if the model supports MP 

As we are gradually starting to build MP-support a cleaner solution will be made in the future, but for now this is good enough to prevent misleading false expectations as reported in https://github.com/huggingface/transformers/issues/9336

(Also for the future, I'm not sure whether it'd be better to check `model.config.architectures`, which would be more precise than checking `model_type` since it's the `architectures` that may or may not support MP within the same `model_type` - but that's a different discussion).

Fixes: #9336

@patrickvonplaten, @sgugger

### CODE REVIEW

File: src/transformers/trainer.py
Comment (patrickvonplaten) [inline src/transformers/trainer.py:247]: Maybe we can check like this for now:

```python
if not hasattr(model, "model_parallel"):
    raise ValueError(f"{model.config.model_type} implementation currently doesn't support model parallelism, therefore --model_parallel cl arg cannot be used")
```

**Review by patrickvonplaten — APPROVED** [review]
Thanks for the PR! I agree that we should add an assert like this.

Also don't think keeping a list we'd manually have to extend is the best way to go here. Maybe just checking whether the model has the attribute `model_parallel` is good enough for now....Wdyt?

**Comment by stas00** [comment]
@alexorona proposed to have this method in PreTrainedModel, which then would break it
https://github.com/huggingface/transformers/pull/9323#issuecomment-752352280

I see this PR as a quick band-aid since we released the new cl arg w/o checking that it always works. And then we will surely improve it as we generalize MP and not leave it this way. This is definitely not how it'll remain in the long run.

**Review by LysandreJik — APPROVED** [review]
LGTM as a hotfix.

Regarding `model_parallel`, if that's to be a method in `PreTrainedModel` and cannot be used to distinguish between models which are parallelizable and models which are not, I think we can go ahead and add a flag for parallelizable models, like `model.parallelizable`. 

Having a way to distinguish between parallelizable models and non-parallelizable models sounds like a must as we continue adding parallelization support.

File: src/transformers/trainer.py
Comment (sgugger) [reply src/transformers/trainer.py:247]: I like that check.

**Review by sgugger — APPROVED** [review]
LGTM with the suggestion from Patrick.

**Comment by stas00** [comment]
So should we merge this one as a hot-fix?

So, an absolute yes to PreTrainedModel.parallelizable accessor - default False, then True for each specific model head that implements it - better than checking arch which doesn't guarantee that it'll have all heads parallelizable. 

And also what do you think about tests? Currently we hardcore a list of parallelizable models - should it remain this way:

https://github.com/huggingface/transformers/blob/086718ac6e20ca2e2cfa3aa0f6da9dc7ee34f6c6/tests/test_modeling_t5.py#L491

or should we automatically derive those from the model by iterating over the heads inside common tests and checking which are parallelizable.

**Comment by sgugger** [comment]
I'd rather merge as a hotfix the proper check and then worry about the tests in a follow up PR (I think we should have a combination of a flag (like for pruning) and checking the models having the attributes there).

**Comment by stas00** [comment]
It no longer will be hot, but yes, I will code that ;) thank you for the feedback, @sgugger

**Comment by sgugger** [comment]
The class `ModelTesterMixin` has a few attributes that control what common tests to apply. I just realized while reading it that it already has the `test_model_parallel` flag so this part is done already. All that is left is just to infer the models to test from the presence of the right attribute :-)

**Comment by stas00** [comment]
OK, I added `model.is_parallelizable` property - let me know if this looks good, or whether you prefer not using a property.

File: src/transformers/modeling_utils.py
Comment (sgugger) [inline src/transformers/modeling_utils.py:407]: A private flag should not appear in the public documentation.

**Review by sgugger — COMMENTED** [review]
I'm fine with this design but it differs from what we were talking about, so we should check the others are fine with it too before merging.

File: src/transformers/modeling_utils.py
Comment (stas00) [reply src/transformers/modeling_utils.py:407]: ah, right! so put the doc for the property then?

that's why it is no longer a hotfix, but it seems to be fine - only one user has filed an issue about using a non-working `--model_parallel` so far.

**Comment by stas00** [comment]
> I'm fine with this design but it differs from what we were talking about, so we should check the others are fine with it too before merging.

Yes, of course.

that's why it is no longer a hotfix, but it seems to be fine - only one user has filed an issue about using a non-working `--model_parallel` so far.

File: src/transformers/modeling_utils.py
Comment (sgugger) [reply src/transformers/modeling_utils.py:407]: Why not have just this class attribute be public and no property?

Comment (stas00) [reply src/transformers/modeling_utils.py]: I thought it was cleaner since it should be read-only, but it's fine as non-property. changed. 

I don't think we have a "way" so that's why I'm never sure when something should be a property or a public attribute.

thank you for the feedback/ideas, @sgugger.

**Comment by stas00** [comment]
So since the only change I proposed is from `parallelizable` to `is_parallelizable`, do you still think we ought to re-validate with @LysandreJik?

File: src/transformers/modeling_utils.py
Comment (sgugger) [reply src/transformers/modeling_utils.py]: TBH `base_model_prefix` is the same, it should be read-only in theory but we have it as a simple class attribute, so let's stay simple for this new attribute too :-)

**Comment by sgugger** [comment]
Yes, let's wait for him to review this tomorrow morning (he's on European time for the next month or so).

**Review by LysandreJik — APPROVED** [review]
LGTM, this is very clean!

**Pull request merged**
